### PR TITLE
ci: Add a workflow to update labels of backported PRs

### LINKED ISSUE #12: Automate setting backport-done labels?
I wonder if it's possible to monitor on merged PRs with label `kind/backport` and then scan through the PR description to locate the set of PRs which have been backported, then automatically update those labels?

### PR DESCRIPTION
**Pull request by pippolo84**
Add a reusable workflow to be called from stable branches whenever a backport PR is merged. The workflow scans the body of the backport PR (that is, a PR like [this one](https://github.com/cilium/cilium/pull/27739)) to get the list of the original PRs, and updates the label `backport-pending/<version>` to `backport-done/<version>` for each of them.

The list of PRs to be updated is fetched from the Backport PR body section labeled `upstream-prs`.

Fixes: https://github.com/cilium/github-actions/issues/12

### CODE REVIEW

**Comment by pippolo84** [comment]
Changed the regexp to list the PRs in order to allow multiple spaces in the tasklist line format:
```diff
diff --git a/.github/workflows/update-label-backport-pr.yaml b/.github/workflows/update-label-backport-pr.yaml
index d6c58bc7ee..0cf981258f 100644
--- a/.github/workflows/update-label-backport-pr.yaml
+++ b/.github/workflows/update-label-backport-pr.yaml
@@ -26,7 +26,7 @@
             env:
                 GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
             run: |
-                echo "${{ inputs.pr-body }}" | grep -Eo '\* \[[x, ]\] #[0-9]+ \(@' | grep -Eo '[0-9]+' | while read -r pr; do
+                echo "${{ inputs.pr-body }}" | grep -Eo '\* +\[[x, ]\] +#[0-9]+ +\(@' | grep -Eo '[0-9]+' | while read -r pr; do
                   echo "Removing label backport-pending/"${{ inputs.branch }}" from pr #${pr}."
                   gh pr edit ${pr} --repo "${GITHUB_REPOSITORY}" --remove-label backport-pending/"${{ inputs.branch }}"
                   echo "Adding label backport-done/"${{ inputs.branch }}" to pr #${pr}."

```

**Review by nebril — APPROVED** [review]
Amazing!

**Review by learnitall — COMMENTED** [review]
Hey @pippolo84, I found five more places where the process of manually setting the done label is referenced. I think these might need to be updated- what do you think?

```
$ cd Documentation && grep -nE '(backport-done|set-labels.py)' -R . -C 2
./contributing/development/reviewers_committers/review_process.rst-124-   .. code-block:: shell-session
./contributing/development/reviewers_committers/review_process.rst-125-
./contributing/development/reviewers_committers/review_process.rst:126:       $ for pr in 26759 24099; do contrib/backporting/set-labels.py $pr done 1.13; done
./contributing/development/reviewers_committers/review_process.rst-127-
./contributing/development/reviewers_committers/review_process.rst-128-Reviewer Teams
--
./contributing/release/backports.rst-168-stable release version X.Y). If any PRs contain labels such as
./contributing/release/backports.rst-169-``backport-pending/X.Y``, ensure that the backport for that PR have been merged
./contributing/release/backports.rst:170:and if so, change the label to ``backport-done/X.Y``.
./contributing/release/backports.rst-171-
./contributing/release/backports.rst-172-Creating the Backports Branch
--
./contributing/release/backports.rst-343-
./contributing/release/backports.rst-344-After the backport PR is merged, if the person who merged the PR didn't take
./contributing/release/backports.rst:345:care of it already, mark all backported PRs with ``backport-done/X.Y`` label
./contributing/release/backports.rst-346-and clear the ``backport-pending/X.Y`` label(s). If the backport pull request
./contributing/release/backports.rst-347-description was generated using the scripts above, then the full command is
--
./contributing/release/backports.rst-350-.. code-block:: shell-session
./contributing/release/backports.rst-351-
./contributing/release/backports.rst:352:   $ GITHUB_TOKEN=xxx for pr in 12589 12568; do contrib/backporting/set-labels.py $pr done 1.8; done
./contributing/release/backports.rst-353-
./contributing/release/backports.rst-354-Backporting Guide for Others
--
./contributing/release/backports.rst-379-.. code-block:: shell-session
./contributing/release/backports.rst-380-
./contributing/release/backports.rst:381:    $ GITHUB_TOKEN=xxx for pr in 12894 12621 12973 12977 12952; do contrib/backporting/set-labels.py $pr done 1.8; done
```

Also, can we delete `set-labels.py`? Or should it be kept around?

**Comment by pippolo84** [comment]
> Hey @pippolo84, I found five more places where the process of manually setting the done label is referenced. I think these might need to be updated- what do you think?

Good catch. :+1: I removed all of those sections, since the workflow will take care of that automatically.

> Also, can we delete `set-labels.py`? Or should it be kept around?

I think the script should stay because it is called by `contrib/backporting/submit-backport` to set the label `backport-pending/<version>`. But since it should not be called anymore to update the label to `backport-done/<version>`, I simplified it removing the `action` parameter and changed the way it is called accordingly.
Also, I fixed the `contrib/backporting/submit-backport` script to take into account the new format of the `v$BRANCH-backport-$(date --rfc-3339=date).txt` file that `contrib/backporting/check-stable` generates.

**Review by learnitall — APPROVED** [review]
Docs look great, thanks!

**Comment by learnitall** [comment]
> Good catch. 👍 I removed all of those sections, since the workflow will take care of that automatically.

Thank you!

> I think the script should stay 

Ah ok makes sense, thanks!

File: .github/workflows/update-label-backport-pr.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          - name: Update labels in backported PRs
+            env:
+              GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+            run: |
+              echo '${{ inputs.pr-body }}' | sed -En "/upstream-prs/ { n; p }" | cut -d ';' -f 1 | grep -Eo '[0-9]+' | while read -r pr; do
+                echo "Removing label backport-pending/${{ inputs.branch }} from pr #${pr}."
+                gh pr edit ${pr} --repo "${GITHUB_REPOSITORY}" --remove-label backport-pending/"${{ inputs.branch }}"
+                echo "Adding label backport-done/${{ inputs.branch }} to pr #${pr}."
+                gh pr edit ${pr} --repo "${GITHUB_REPOSITORY}" --add-label backport-done/"${{ inputs.branch }}"
+              done || true
```
Comment (aanm) [inline .github/workflows/update-label-backport-pr.yaml]: Remove `|| true` I don't think it's needed. Maybe we just need to make sure that the workflow is not failing.

**Review by aanm — CHANGES_REQUESTED** [review]

**Review by viktor-kurchenko — APPROVED** [review]

File: .github/workflows/update-label-backport-pr.yaml (hunk 1 above)
Comment (pippolo84) [reply .github/workflows/update-label-backport-pr.yaml]: Removed and added this to the docs:

```diff
+After the Backports are Merged
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After the backport PR is merged, the GH workflow "Call Backport Label Updater"
+should take care of marking all backported PRs with the ``backport-done/X.Y``
+label and clear the ``backport-pending/X.Y`` label(s).
+Verify that the workflow succeeded by looking `here
+<https://github.com/cilium/cilium/actions/workflows/call-backport-label-updater.yaml>`_.
```

**Review by aanm — APPROVED** [review]

**Comment by pippolo84** [comment]
This PR adds a new workflow and changes the contributing documentation. Running the full CI is pointless, so I'm marking this ready-to-merge

File: contrib/backporting/submit-backport [hunk 2]
```diff
@@ -59,12 +59,12 @@ else
...
 
-prs=$(grep "contrib/backporting/set-labels.py" $SUMMARY | sed -e 's/^.*for pr in \([0-9 ]\+\);.*$/\1/g')
+prs=$(sed -En "/upstream-prs/ { n; p }" < "$SUMMARY")
 echo -e "\nUpdating labels for PRs $prs\n" 2>&1
 echo -n "Set labels for all PRs above? [Y/n] "
 read set_all_labels
 if [ "$set_all_labels" != "n" ]; then
     for pr in $prs; do
-        $DIR/set-labels.py $pr pending $BRANCH;
+        $DIR/set-labels.py $pr $BRANCH;
```
Comment (joestringer) [inline contrib/backporting/submit-backport:68]: This change broke the script. I've sent a fix: https://github.com/cilium/cilium/pull/33826 .

(I realize a lot of folks use other tooling nowadays, but this is the tooling in the tree, so it's good to know it works).

**Pull request merged**